### PR TITLE
HTML debugging view

### PR DESCRIPTION
**Pull request by iamdanfox**
## Before this PR

I found it hard to imagine how all of the abstractions in this project fit together, which made iterating on the actual logic a bit confusing.

## After this PR

==COMMIT_MSG==
A HTML debugging view can be generated to help iterate on the formatter logic
==COMMIT_MSG==

Shows how `Op`s are laid out, before building them into `Levels`:

image

Shows the different paths the formatter took to reach the output, and allows exploring what the various "explorations" from each level would have produced if selected:

image

formatter-exploration-2

Before merging:
- [x] don't run this during normal formatting

### CODE REVIEW

File: palantir-java-format/build.gradle [hunk 1]
```diff
@@ -10,6 +10,9 @@ dependencies {
     api 'com.google.guava:guava'
     implementation 'com.google.errorprone:javac-shaded'
     implementation 'org.functionaljava:functionaljava'
+    implementation 'com.fasterxml.jackson.core:jackson-databind' // purely for the debugger
+    implementation 'com.fasterxml.jackson.datatype:jackson-datatype-jdk8'
+    implementation 'com.fasterxml.jackson.module:jackson-module-parameter-names'
```
Comment (iamdanfox) [inline palantir-java-format/build.gradle:15]: @dansanduleac I think I'd like to keep the dependencies of palantir-java-format as minimal as possible - thoughts on keeping these in the `:debugger` project and taking a compileOnly dependency on that project?

File: palantir-java-format/src/main/java/com/palantir/javaformat/doc/Obs.java [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+ */
+
+package com.palantir.javaformat.doc;
+
+import java.util.Optional;
+import java.util.OptionalInt;
+import java.util.function.Function;
+import javax.annotation.CheckReturnValue;
+
+public interface Obs {
```
Comment (iamdanfox) [inline palantir-java-format/src/main/java/com/palantir/javaformat/doc/Obs.java]: Think we gotta tell people a bit of a story about what's going on here.

**Comment by ferozco** [comment]
@dansanduleac could you push up a gif of your latest work? Very interested to see this in action

**Comment by dansanduleac** [comment]
👍

**Comment by iamdanfox** [comment]
:+1:

**Pull request merged**